src/Interpreters/ProcessList.h

```diff
@@ -310,6 +310,14 @@ class ProcessList
     mutable std::mutex mutex;
     mutable std::condition_variable have_space;        /// Number of currently running queries has become less than maximum.
 
+    struct LockAndBlocker
+    {
+        std::lock_guard<std::mutex> guard;
+        OvercommitTrackerBlockerInThread blocker;
+    };
+
+    LockAndBlocker safeLock() const noexcept { return { std::lock_guard{mutex}, {} }; }
```

Comment:
It's still possible to accidentally lock the mutex without calling this method. Maybe we should move `mutex` to a base class and make it private or something.